files/virt_volume.sh

```diff
@@ -17,82 +17,119 @@
 # Ensure that a libvirt volume exists, optionally uploading an image.
 # On success, output a JSON object with a 'changed' item.
 
-if [[ $# -ne 4 ]] && [[ $# -ne 5 ]]; then
-    echo "Usage: $0 <name> <pool> <capacity> <format> [<image>]"
+# Parse options
+OPTIND=1
+
+if [[ $# -gt 4 ]] && [[ $# -lt 6 ]]; then
+    echo "Usage: $0 -n <name> -p <pool> -c <capacity> -f <format> [-i <source image> | -b <backing image>]"
     exit 1
 fi
 
-NAME=$1
-POOL=$2
-CAPACITY=$3
-FORMAT=$4
-IMAGE=$5
+while getopts ":n:p:c:f:i:b:" opt; do
+    case ${opt} in
+        n) NAME=$OPTARG;;
+        p) POOL=$OPTARG;;
+        c) CAPACITY=$OPTARG;;
+        f) FORMAT=$OPTARG;;
+        i) IMAGE=$OPTARG;;
+        b) BACKING_IMAGE=$OPTARG;;
+        \?)
+            echo "Invalid option: -$OPTARG" >&2
+            exit 1
+            ;;
+        :)
+            echo "Option -$OPTARG requires an argument." >&2
+            exit 1
+            ;;
+    esac
+done
+
+if [[ -n $IMAGE && -n $BACKING_IMAGE ]]; then
+  echo "Options -i and -b are mutually exclusive." >&2
+  exit 1
+fi
 
 # Check whether a volume with this name exists.
-output=$(virsh vol-info --pool $POOL --vol $NAME 2>&1)
+output=$(virsh vol-info --pool "$POOL" --vol "$NAME" 2>&1)
 result=$?
 if [[ $result -eq 0 ]]; then
     echo '{"changed": false}'
     exit 0
 elif ! echo "$output" | grep 'Storage volume not found' >/dev/null 2>&1; then
-    echo "Unexpected error while getting volume info"
+    echo "Unexpected error while getting volume info" >&2
     echo "$output"
     exit $result
 fi
 
 # Create the volume.
-output=$(virsh vol-create-as --pool $POOL --name $NAME --capacity $CAPACITY --format $FORMAT 2>&1)
-result=$?
+if [[ -n $BACKING_IMAGE ]]; then
+    if [[ "$FORMAT" -ne "qcow2" ]]; then
```

Comment:
`-ne` is for numbers, use `!=` for strings.

Reply:
> `-ne` is for numbers, use `!=` for strings.

Thanks for cleaning up my somewhat sloppy test there.